### PR TITLE
[FEATURE] Enables offline /score for embedding models

### PR DESCRIPTION
**Pull request by gmarinho2**
Enables LLM.score() for all embedding models. The request_id consists of the request_ids of each embedding in the pair, separated by "_". The prompt_token_ids are the concatenation of all the token ids, in order and separated by the padding token when it is available. This PR is the first of two for completing the issue. The second PR will implement the same feature in the OpenAI API.

Issue: [[Feature]: Enable /score endpoint for all embedding models (1/2)](https://github.com/vllm-project/vllm/issues/10752)

### CODE REVIEW

File: vllm/entrypoints/llm.py [hunk 1]
```diff
@@ -1045,37 +1037,92 @@ def ensure_str(prompt: SingletonPrompt):
         if len(text_1) == 1:
             text_1 = text_1 * len(text_2)
 
-        input_pairs = [(t1, t2) for t1, t2 in zip(text_1, text_2)]
-        pooling_params = PoolingParams()
+        if self.llm_engine.model_config.task != "score":
```
Comment (maxdebayser) [inline vllm/entrypoints/llm.py:1040]:
```suggestion
if self.llm_engine.model_config.task not in ["embed", "score"]:
```

File: vllm/entrypoints/llm.py [hunk 2]
```diff
@@ -1045,37 +1037,92 @@ def ensure_str(prompt: SingletonPrompt):
         if len(text_1) == 1:
```
Comment (maxdebayser) [inline vllm/entrypoints/llm.py:1037]: We only want do this for cross encoder models. For embedding models where each prompt is sent on it's own, we don't want to send the prompt N times.

File: vllm/entrypoints/llm.py [hunk 3]
```diff
@@ -1045,37 +1037,92 @@ def ensure_str(prompt: SingletonPrompt):
...
+            return [ScoringRequestOutput.from_base(item) for item in items]
+
+        elif self.llm_engine.model_config.has_pooling:
+            text = text_1 + text_2
+            encoded_text = self.encode(text)
+
+            encoded_text_1 = encoded_text[0:len(text_1)]
+            encoded_text_2 = encoded_text[len(text_1):]
+
+            input_pairs_2 = [(t1, t2)
```
Comment (maxdebayser) [inline vllm/entrypoints/llm.py:1088]: See comment above on replicating the same sequence. Basically we want to move that logic from input to output. Also, can we change the variable name to output_pairs or something like that?

File: vllm/entrypoints/llm.py [hunk 4]
```diff
@@ -1045,37 +1037,92 @@ def ensure_str(prompt: SingletonPrompt):
...
+                             for t1, t2 in zip(encoded_text_1, encoded_text_2)]
+
+            scores = []
+            cosSim = torch.nn.CosineSimilarity(0)
+
+            if tokenizer.pad_token is None:
+                for token_1, token_2 in input_pairs_2:
```
Comment (maxdebayser) [inline vllm/entrypoints/llm.py:1095]: Here the items aren't tokens but embedding vectors. Let's change the variable name for more clarity.

File: vllm/entrypoints/llm.py [hunk 5]
```diff
@@ -1045,37 +1037,92 @@ def ensure_str(prompt: SingletonPrompt):
...
+            encoded_text_1 = encoded_text[0:len(text_1)]
+            encoded_text_2 = encoded_text[len(text_1):]
+
+            input_pairs_2 = [(t1, t2)
+                             for t1, t2 in zip(encoded_text_1, encoded_text_2)]
+
+            scores = []
+            cosSim = torch.nn.CosineSimilarity(0)
+
+            if tokenizer.pad_token is None:
```
Comment (maxdebayser) [inline vllm/entrypoints/llm.py:1094]: We can resolve the mypy problem testing with `getattr`.
```suggestion
if getattr(tokenizer, "pad_token", None) is None:
```

File: vllm/entrypoints/llm.py (hunk 1 above)
Comment (joerunde) [reply vllm/entrypoints/llm.py:1040]: It'd be nice to move this up to the top of the method as well, before other work is done

File: vllm/entrypoints/llm.py [hunk 6]
```diff
@@ -1045,37 +1037,92 @@ def ensure_str(prompt: SingletonPrompt):
...
-        items = self.engine_class.validate_outputs(outputs,
-                                                   PoolingRequestOutput)
+            outputs = self._run_engine(use_tqdm=use_tqdm)
+            items = self.engine_class.validate_outputs(outputs,
+                                                       PoolingRequestOutput)
 
-        return [ScoringRequestOutput.from_base(item) for item in items]
+            return [ScoringRequestOutput.from_base(item) for item in items]
+
+        elif self.llm_engine.model_config.has_pooling:
```
Comment (maxdebayser) [inline vllm/entrypoints/llm.py:1081]:
```suggestion
elif self.llm_engine.model_config.runner_type == "pooling":
```

**Comment by joerunde** [comment]
@maxdebayser @gmarinho2 This looks like it only touches the offline entrypoint, but the PR title mentions `/score endpoint` which I had assumed meant the online REST interface.

It's not 100% clear to me what was intended- is there more work planned to support the online interface or are we only aiming for offline?

**Comment by maxdebayser** [comment]
@joerunde, we're aiming for both. @gmarinho2 started with the offline API first.

File: vllm/entrypoints/llm.py [hunk 7]
```diff
@@ -1042,40 +1039,87 @@ def ensure_str(prompt: SingletonPrompt):
...
+                if getattr(tokenizer, "pad_token", None) is None:
+                    tokens = embed_1.prompt_token_ids + embed_2.prompt_token_ids
+                else:
+                    tokens = embed_1.prompt_token_ids + [
+                        tokenizer.pad_token_type_id
+                    ] + embed_2.prompt_token_ids
+
+                scores.append(
+                    PoolingRequestOutput(
+                        request_id="unk",  #unk ou id incremental?
```
Comment (joerunde) [inline vllm/entrypoints/llm.py:1112]: Some other ideas:

- Use the first request ID from the batch: `encoded_text[0].request_id`
- Use the first and last request ID from the batch `f"{encoded_text[0].request_id}_{encoded_text[-1].request_id}"` 
- Go super verbose and put all of them: `"_".join(x.request_id for x in encoded_text)`

File: vllm/entrypoints/llm.py [hunk 8]
```diff
@@ -1042,40 +1039,87 @@ def ensure_str(prompt: SingletonPrompt):
...
-        input_pairs = [(t1, t2) for t1, t2 in zip(text_1, text_2)]
-        pooling_params = PoolingParams()
+            input_pairs = [(t1, t2)
+                           for t1, t2 in zip(text_1, text_2)]  #output_pairs?
 
-        tokenization_kwargs: Dict[str, Any] = {}
-        if truncate_prompt_tokens is not None:
-            tokenization_kwargs["truncation"] = True
-            tokenization_kwargs["max_length"] = truncate_prompt_tokens
+            if isinstance(tokenizer, MistralTokenizer):
```
Comment (maxdebayser) [inline vllm/entrypoints/llm.py:1049]: Let's move this higher up, right after:
````python
if self.llm_engine.model_config.is_cross_encoder:
```

File: vllm/entrypoints/llm.py (hunk 7 above)
Comment (maxdebayser) [reply vllm/entrypoints/llm.py:1112]: Since we have the IDs for all embedding results, we could grab the ids and concatenate them for each scoring output:
```python
                pair_id  = f"{embed_1.request_id}_{embed_2.request_id}"
```

File: vllm/entrypoints/llm.py [hunk 9]
```diff
@@ -1042,40 +1039,87 @@ def ensure_str(prompt: SingletonPrompt):
...
+                encoded_text_1 = encoded_text_1 * len(encoded_text_2)
+
+            output_pairs = [(t1, t2)
+                            for t1, t2 in zip(encoded_text_1, encoded_text_2)]
+
+            scores = []
+            cosSim = torch.nn.CosineSimilarity(0)
+
+            for embed_1, embed_2 in output_pairs:
+                pairs_score = cosSim(embed_1.outputs.data,
```
Comment (maxdebayser) [inline vllm/entrypoints/llm.py:1100]: Can we rename this to `pair_score` to make it clear that it refers to a single pair?

File: tests/models/embedding/language/test_scoring.py [hunk 10]
```diff
@@ -87,3 +93,101 @@ def test_llm_N_to_N(vllm_runner, hf_runner, model_name, dtype: str):
...
+
+    with hf_runner(emb_model_name, dtype=dtype,
+                   is_sentence_transformer=True) as hf_model:
+        hf_embeddings = [
+            hf_model.encode(text_pair) for text_pair in text_pairs
+        ]
+        hf_outputs = [
+            F.cosine_similarity(torch.tensor(encoded_pair[0]),
+                                torch.tensor(encoded_pair[1]),
+                                dim=0) for encoded_pair in hf_embeddings
```
Comment (maxdebayser) [inline tests/models/embedding/language/test_scoring.py:147]:
````suggestion
            F.cosine_similarity(*map(torch.tensor, pair),
                                dim=0) for pair in hf_embeddings
```

File: tests/models/embedding/language/test_scoring.py [hunk 11]
```diff
@@ -87,3 +93,101 @@ def test_llm_N_to_N(vllm_runner, hf_runner, model_name, dtype: str):
...
```
Comment (maxdebayser) [inline tests/models/embedding/language/test_scoring.py:180]:
```suggestion
            F.cosine_similarity(*map(torch.tensor, pair),
                                dim=0) for pair in hf_embeddings
```

File: tests/models/embedding/language/test_scoring.py [hunk 12]
```diff
@@ -87,3 +93,101 @@ def test_llm_N_to_N(vllm_runner, hf_runner, model_name, dtype: str):
...
+
+    text_pair = [TEXTS_1[0], TEXTS_2[0]]
+
+    with hf_runner(emb_model_name, dtype=dtype,
+                   is_sentence_transformer=True) as hf_model:
+        hf_embeddings = hf_model.encode(text_pair)
+        hf_outputs = [
+            F.cosine_similarity(torch.tensor(hf_embeddings[0]),
+                                torch.tensor(hf_embeddings[1]),
+                                dim=0)
```
Comment (maxdebayser) [inline tests/models/embedding/language/test_scoring.py:115]:
```suggestion
            F.cosine_similarity(*map(torch.tensor, hf_embeddings),
                                 dim=0)
```

**Review by maxdebayser — APPROVED** [review]
I've left some suggestions, but it looks good to me. I think we can open this as a PR now.

File: vllm/entrypoints/llm.py [hunk 13]
```diff
@@ -1047,25 +1049,20 @@ def score(
...
 
-        if not self.llm_engine.model_config.is_cross_encoder:
-            raise ValueError("Your model does not support cross encoding")
-        if self.llm_engine.model_config.task != "score":
-            raise ValueError("Score API is only enabled for `--task score`")
-
-        tokenizer = self.llm_engine.get_tokenizer()
-
-        if isinstance(tokenizer, MistralTokenizer):
+        if self.llm_engine.model_config.task not in ["embed", "score"]:
```
Comment (DarkLight1337) [inline vllm/entrypoints/llm.py:1052]:
```suggestion
        if self.llm_engine.model_config.task not in ("embed", "score"):
```

File: vllm/entrypoints/llm.py [hunk 14]
```diff
@@ -1091,40 +1088,84 @@ def ensure_str(prompt: SingletonPrompt):
...
 
-        outputs = self._run_engine(use_tqdm=use_tqdm)
-        items = self.engine_class.validate_outputs(outputs,
-                                                   PoolingRequestOutput)
+            return [ScoringRequestOutput.from_base(item) for item in items]
+
+        elif self.llm_engine.model_config.runner_type == "pooling":
+            encoded_text = self.encode(text_1 + text_2)
+            encoded_text_1 = encoded_text[0:len(text_1)]
+            encoded_text_2 = encoded_text[len(text_1):]
```
Comment (DarkLight1337) [inline vllm/entrypoints/llm.py:1135]: We should base this on the length of the token IDs rather than text.

File: vllm/entrypoints/llm.py [hunk 15]
```diff
@@ -1091,40 +1088,84 @@ def ensure_str(prompt: SingletonPrompt):
...
+                encoded_text_1 = encoded_text_1 * len(encoded_text_2)
+
+            output_pairs = [(t1, t2)
+                            for t1, t2 in zip(encoded_text_1, encoded_text_2)]
+
+            scores = []
+            cosSim = torch.nn.CosineSimilarity(0)
+
+            for embed_1, embed_2 in output_pairs:
+                pair_score = cosSim(embed_1.outputs.data, embed_2.outputs.data)
```
Comment (DarkLight1337) [inline vllm/entrypoints/llm.py:1147]:
```suggestion
            scorer = torch.nn.CosineSimilarity(0)

            for embed_1, embed_2 in output_pairs:
                pair_score = scorer(embed_1.outputs.data, embed_2.outputs.data)
```

Use snake case, not camel case

**Review by DarkLight1337 — COMMENTED** [review]
Some initial comments.

I also suggest splitting out the logic for scoring and general embedding models into separate functions.

File: vllm/entrypoints/llm.py (hunk 14 above)
Comment (gmarinho2) [reply vllm/entrypoints/llm.py:1135]: In this context text_1 and text_2 are already lists, so the length being used is actually from these lists and not from actual text. Maybe a new variable `texts_list = text_1 + text_2` would make it more clear?

Comment (DarkLight1337) [reply vllm/entrypoints/llm.py:1135]: Oh, I see. In that case there is no need to change that (I didn't look at the function signature)

File: vllm/entrypoints/llm.py [hunk 16]
```diff
@@ -1091,40 +1088,84 @@ def ensure_str(prompt: SingletonPrompt):
...
-                                                   PoolingRequestOutput)
+            return [ScoringRequestOutput.from_base(item) for item in items]
+
+        elif self.llm_engine.model_config.runner_type == "pooling":
+            encoded_text = self.encode(text_1 + text_2)
+            encoded_text_1 = encoded_text[0:len(text_1)]
+            encoded_text_2 = encoded_text[len(text_1):]
+
+            if len(encoded_text_1) == 1:
+                encoded_text_1 = encoded_text_1 * len(encoded_text_2)
```
Comment (DarkLight1337) [inline vllm/entrypoints/llm.py:1138]: What if `len(encoded_text_2) == 1` but `len(encoded_text_1) > 1`?

File: vllm/entrypoints/llm.py (hunk 14 above)
Comment (DarkLight1337) [reply vllm/entrypoints/llm.py:1135]: Actually, perhaps it would be better to rename this to `encode_outputs`? Since each item is actually an object rather than the embedding vector itself.

Comment (gmarinho2) [reply vllm/entrypoints/llm.py:1135]: I see, it is in fact a better name.

File: vllm/entrypoints/llm.py (hunk 16 above)
Comment (gmarinho2) [reply vllm/entrypoints/llm.py:1138]: This cannot happen since the inputs allowed are only 1:1, 1:N or N:N. This is guaranteed by the conditional:

`if len(text_1) > 1 and len(text_1) != len(text_2):
            raise ValueError("Input lengths must be either 1:1, 1:N or N:N")`

Comment (DarkLight1337) [reply vllm/entrypoints/llm.py]: Got it.

**Comment by DarkLight1337** [comment]
> I also suggest splitting out the logic for scoring and general embedding models into separate functions.

Have you addressed this?

File: vllm/entrypoints/llm.py [hunk 17]
```diff
@@ -996,6 +997,90 @@ def classify(
 
         return [ClassificationRequestOutput.from_base(item) for item in items]
 
+    def _embedding_score(self, tokenizer, text_1: List[str],
+                         text_2: List[str]) -> List[ScoringRequestOutput]:
+
```
Comment (maxdebayser) [inline vllm/entrypoints/llm.py:1002]: Can you make the signature of this function the same as the `_cross_encoding_score`? 

```suggestion
    def _embedding_score(
            self, tokenizer, truncate_prompt_tokens,
            text_1: List[str | TextPrompt | TokensPrompt],
            text_2: List[str | TextPrompt | TokensPrompt], use_tqdm,
            lora_request,
            prompt_adapter_request) -> List[ScoringRequestOutput]:
```

The `use_tqdm`, `lora_request`, `prompt_adapter_request` parameters should be forwared to `self.encode()`. `tokenizer` and `truncate_prompt_tokens` won't be used for now but I want to open a PR adding support for `truncate_prompt_tokens` in `LLM.embed()` and `LLM.encode()`

File: vllm/entrypoints/llm.py [hunk 18]
```diff
@@ -996,6 +997,97 @@ def classify(
 
         return [ClassificationRequestOutput.from_base(item) for item in items]
 
+    def _embedding_score(self, tokenizer, truncate_prompt_tokens,
+                         text_1: List[Union[str, TextPrompt, TokensPrompt]],
+                         text_2: List[Union[str, TextPrompt, TokensPrompt]],
+                         use_tqdm, lora_request,
+                         prompt_adapter_request) -> List[ScoringRequestOutput]:
```
Comment (DarkLight1337) [inline vllm/entrypoints/llm.py:1004]: Please add type annotations for the parameters.

File: vllm/entrypoints/llm.py [hunk 19]
```diff
@@ -1047,25 +1149,20 @@ def score(
...
-        if self.llm_engine.model_config.task != "score":
-            raise ValueError("Score API is only enabled for `--task score`")
-
-        tokenizer = self.llm_engine.get_tokenizer()
-
-        if isinstance(tokenizer, MistralTokenizer):
+        if self.llm_engine.model_config.task not in ("embed", "score"):
             raise ValueError(
-                "MistralTokenizer not supported for cross-encoding")
+                "Score API is only enabled for `--task embed or score`")
```
Comment (DarkLight1337) [inline vllm/entrypoints/llm.py:1154]:
```suggestion
                "Score API is only enabled for `--task embed` or `--task score`")
```

File: vllm/entrypoints/llm.py [hunk 20]
```diff
@@ -1032,6 +1133,7 @@ def score(
             A list of ``ScoringRequestOutput`` objects containing the
             generated scores in the same order as the input prompts.
         """
+
```
Comment (DarkLight1337) [inline vllm/entrypoints/llm.py:1136]:
```suggestion
```
Avoid unnecessary line changes

**Comment by joerunde** [comment]
Added the `ready` tag so the full CI run can run your tests as well. Looks like the only fastcheck failure was a network issue

File: vllm/entrypoints/llm.py (hunk 19 above)
Comment (DarkLight1337) [reply vllm/entrypoints/llm.py]: It seems that you forgot to address this and the next comment.

Comment (gmarinho2) [reply vllm/entrypoints/llm.py]: I made these changes in this commit: https://github.com/vllm-project/vllm/pull/12021/commits/f8b8d8c80d324df4cf228d251591aa6bc22f4451

File: vllm/entrypoints/llm.py (hunk 20 above)
Comment (DarkLight1337) [reply vllm/entrypoints/llm.py]: I still see this line in the PR.

**Review by DarkLight1337 — APPROVED** [review]
Sorry for the confusion, looks good now.
